As system, user's tap time is stored

iamdejan opened the pull request:
What I've done:
1. System can store user's time in memory until the room is deleted.
2. System can find the winner after all user register tap times.
3. System publishes `GAME_END` event.

![Screen Shot 2019-07-11 at 15 44 10](https://user-images.githubusercontent.com/15686093/61036346-2551a980-a3f3-11e9-884b-cccfff207a47.png)

iamdejan committed:
Refactor one-letter variable

iamdejan committed:
Add tap time to event payload

iamdejan committed:
Revise requests

iamdejan committed:
Implement register tap time feature

iamdejan committed:
Implement GAME_END event

iamdejan committed:
FInish GAME_END implementation

iamdejan committed:
Remove unecessary type-casting

iamdejan committed:
Adding delete member from room feature

iamdejan committed:
Remove MaximumJoinDurationInSeconds()

I think Godot won't need this.

nieltg reviewed (changes requested):
Overall, this is much better. One pitfall that you forgot most is reducing public members. (either remove it or just don't make unused public members)

**Note:** Many public members mean you have to maintenance more things since anyone can depend on those public members. Changing private members is more easy than changing public members since you know that only your package that depends on those private members.

### About `Payload`

I intended `Payload` to be a constructor parameter actually. Look at `RoomPayload` and `MemberPayload`.

To create a `Room`, you need:
- `MaxMemberCount`, so the room may starts after the room is full.

To create a `Member`, you need:
- `ID`, the member ID.
- `Name`, the display name of the member.

That's why I created `Payload` things. But, more question. Why do I put these on a `struct`? Why don't just write it as parameters of the constructors?

It's because `struct` is more explicit when stating parameters. You have to type the key, then the value of the parameter. It's not like Python when you can call things like this: `func1(name="hello", value=5)`. In Go, you have to put it on the `struct`.

So, how to clean up this mess? We have to restate our design here. I'm inviting you to also think about this and I'll review your opinion also while I'm thinking for this.

nieltg committed:
pkg/iohttp/util: Rename v -> value

nieltg committed:
pkg/iohttp: Rename id -> roomID, user_id -> memberID

nieltg committed:
pkg/iohttp: Use DELETE method instead of GET

nieltg committed:
pkg: DeleteMember returns bool instead oe Member

nieltg committed:
pkg/iohttp/handler: Rename error -> err

error is Go lang. built-in interface

iamdejan committed:
Update room.go

Changes:
1) Lock mechanism at create & delete member, not at increment / decrement process.
2) Put constant number into constant variable.

iamdejan committed:
Change constant into Golang defined constant

iamdejan committed:
Delete room after game is finished
